### PR TITLE
Fix #1416: add WebAssembly lexer

### LINKED ISSUE #1416: Support for WebAssembly
Would be nice to have.

There are VS Code Syntax at https://github.com/wasmerio/vscode-wasm/blob/master/syntaxes/wat.json.

### PR DESCRIPTION
**Pull request by jendrikw**
This is the WebAssembly lexer I've been using and it worked quite well.

There is one test failure that I'm not sure how to resolve:

```
/path/pygments/.tox/lint/lib/python3.8/site-packages/pygments/lexers/webassembly.py:252: (WatLexer:arguments:pat#6) E999: Matches empty string
              (r'', Text, '#pop'),
                           ^ here
```

Fixes #1416.

### CODE REVIEW

**Comment by birkenfeld** [comment]
For the empty case, use `default('#pop')`.

**Review by birkenfeld — CHANGES_REQUESTED** [review]
Thanks for the PR! A few points still need addressing...

File: pygments/lexers/webassembly.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+f64.convert_i32_s
+f64.convert_i32_u
+f64.convert_i64_s
+f64.convert_i64_u
+f64.promote_f32
+i32.reinterpret_f32
+i64.reinterpret_f64
+f32.reinterpret_i32
+f64.reinterpret_i64
+""".split()
```
[the remark this answers is not in the document]
Comment (jendrikw) [reply pygments/lexers/webassembly.py:200]: I'm not sure what you mean by them? Should I do them like the keywords?

[the remark this answers is not in the document]
Comment (birkenfeld) [reply pygments/lexers/webassembly.py:200]: yes, please, but as a tuple (with `(...)`). This compiles the whole list into a constant without runtime overhead.

File: pygments/lexers/webassembly.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+class WatLexer(RegexLexer):
+    name = 'WebAssembly'
+    aliases = ['wast', 'wat']
+    filenames = ['*.wat', '*.wast']
+
+    tokens = {
+        'root': [
+            (words(keywords, suffix=r'(?=[^a-z_\.])'), Keyword),
+            (words(builtins), Name.Builtin, 'arguments'),
+            (r'i32.const', Name.Builtin),
```
[the remark this answers is not in the document]
Comment (jendrikw) [reply pygments/lexers/webassembly.py:212]: I can't remember why I did this, can be removed.

File: pygments/lexers/webassembly.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+            (r'[+-]?0x[\dA-Fa-f](_?[\dA-Fa-f])*', Number.Hex),
+            (r'[+-]?\d(_?\d)*', Number.Integer),
+            (r'[\(\)]', Punctuation),
+            (r'"', String.Double, 'string'),
+            (r'\s+', Text),
+        ],
+        'nesting_comment': [
+            (r'\(;', Comment.Multiline, '#push'),
+            (r';\)', Comment.Multiline, '#pop'),
+            (r'(.|\n)', Comment.Multiline)
```
[the remark this answers is not in the document]
Comment (jendrikw) [reply pygments/lexers/webassembly.py:231]: I couldn't quite get this to work with nesting comments, but this does the trick:
```
        'nesting_comment': [
            (r'\(;', Comment.Multiline, '#push'),
            (r';\)', Comment.Multiline, '#pop'),
            (r'(\([^;]|;[^\)]|[^;\(])+', Comment.Multiline),
        ],
```

File: pygments/lexers/webassembly.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        ],
+        'string': [
+            (r'\\[\dA-Fa-f][\dA-Fa-f]*', String.Escape),
+            (r'\t', String.Escape),
```
[the remark this answers is not in the document]
Comment (jendrikw) [reply pygments/lexers/webassembly.py:235]: Thanks, I tend to get confused with all the escaping.

**Comment by jendrikw** [comment]
I think i addressed all comments.

File: pygments/lexers/webassembly.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (birkenfeld) [reply pygments/lexers/webassembly.py]: Ah right, I missed that comments nest. In that case it's
```
(r'[^;(]+', Comment.Multiline),
(r'[;(]', Comment.Multiline),
```

[the remark this answers is not in the document]
Comment (jendrikw) [reply pygments/lexers/webassembly.py]: Thanks, I changed it to your suggestion.

**Comment by Anteru** [comment]
@birkenfeld Could you please take another look at this one?

**Comment by birkenfeld** [comment]
The tests need to be fixed up (output file for the example, and snippets instead of the .py file)... Otherwise looks good, just the "versionadded" is now 2.8

**Comment by Anteru** [comment]
Let's make this 2.9 as the target. Given this needs the tests updated, and 2.8 coming out today, no need to rush this in.

File: pygments/lexers/webassembly.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    pygments.lexers.webassembly
+    ~~~~~~~~~~~~~~~~~~~
+
+    Lexers for the WebAssembly text format.
+
+    The grammar can be found at https://github.com/WebAssembly/spec/blob/master/interpreter/README.md
+    and https://webassembly.github.io/spec/core/text/.
+
+
+    :copyright: Copyright 2006-2020 by the Pygments team, see AUTHORS.
```
Comment (Mesteery) [inline pygments/lexers/webassembly.py:12]:
```suggestion
    :copyright: Copyright 2006-2021 by the Pygments team, see AUTHORS.
```

File: tests/test_webassembly.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
+# -*- coding: utf-8 -*-
+"""
+    Basic WatLexer Test
+    ~~~~~~~~~~~~~~~~~~~~
+
+    :copyright: Copyright 2006-2020 by the Pygments team, see AUTHORS.
```
Comment (Mesteery) [inline tests/test_webassembly.py:6]:
```suggestion
    :copyright: Copyright 2006-2021 by the Pygments team, see AUTHORS.
```

**Review by Mesteery — CHANGES_REQUESTED** [review]

**Comment by Anteru** [comment]
Can you please merge the latest master into this and update for the new test system? That seems to be causing the failures, and in theory all you need to do is to move your test file into a folder and add a golden test output file.

**Comment by PhilippGackstatter** [comment]
Hi @jendrikw, I would love to have `wat` highlighting and was wondering if or when you might find the time to finish this PR? Thanks a lot!

**Comment by jendrikw** [comment]
I thinks this is what you hand in mind. Could you please check if this can be merged?

**Comment by Anteru** [comment]
Pretty much yes -- nearly there :) Your unit test file should also use the new mechanism, given you only test an input for a given set of tokens. Look at tests/snippets -- you can extract the input per test into a single file, and let pytest take care of the tokens. That's easier for us because we can regenerate those without having to touch Python code.

**Comment by jendrikw** [comment]
That should be it.

**Comment by Anteru** [comment]
Merged, thanks!

**Pull request merged**